src/serde.rs

```diff
@@ -0,0 +1,269 @@
+        state.serialize_field("errors_on_transmitted", &self.errors_on_transmitted())?;
+        state.serialize_field(
+            "total_errors_on_transmitted",
+            &self.total_errors_on_transmitted(),
+        )?;
+        state.serialize_field("mac_address", &self.mac_address())?;
+
+        state.end()
+    }
+}
+
+impl Serialize for dyn ComponentExt {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("Component", 4)?;
+
+        state.serialize_field("temperature", &self.temperature())?;
+        state.serialize_field("max", &self.max())?;
+        state.serialize_field("critical", &self.critical())?;
+        state.serialize_field("label", &self.label())?;
+
+        state.end()
+    }
+}
+
+impl Serialize for dyn UserExt {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("User", 1)?;
+
+        let uid = *self.id().deref();
+        state.serialize_field("id", &uid)?;
+
+        let gid = *self.group_id().deref();
+        state.serialize_field("group_id", &gid)?;
+
+        state.serialize_field("name", &self.name())?;
+        state.serialize_field("groups", &self.groups())?;
+
+        state.end()
+    }
+}
+
+use crate::{DiskType, DiskUsage, MacAddr, ProcessStatus};
+
+impl Serialize for DiskType {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let (index, variant, maybe_value) = match *self {
+            DiskType::HDD => (0, "HDD", None),
+            DiskType::SSD => (1, "SSD", None),
+            DiskType::Unknown(ref s) => (2, "Unknown", Some(s)),
+        };
+
+        if let Some(ref value) = maybe_value {
+            serializer.serialize_newtype_variant("DiskType", index, variant, value)
+        } else {
+            serializer.serialize_unit_variant("DiskType", index, variant)
+        }
+    }
+}
+
+impl Serialize for ProcessStatus {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let (index, variant, maybe_value) = match *self {
+            ProcessStatus::Idle => (0, "Idle", None),
+            ProcessStatus::Run => (1, "Run", None),
+            ProcessStatus::Sleep => (2, "Sleep", None),
+            ProcessStatus::Stop => (3, "Stop", None),
+            ProcessStatus::Zombie => (4, "Zombie", None),
+            ProcessStatus::Tracing => (5, "Tracing", None),
+            ProcessStatus::Dead => (6, "Dead", None),
+            ProcessStatus::Wakekill => (7, "Wakekill", None),
+            ProcessStatus::Waking => (8, "Waking", None),
+            ProcessStatus::Parked => (9, "Parked", None),
+            ProcessStatus::LockBlocked => (10, "LockBlocked", None),
+            ProcessStatus::Unknown(n) => (11, "Unknown", Some(n)),
+        };
```

Comment:
Two things about this one: why do you want an index? Why do you want the `Unknown` value? I really think the first one isn't needed so with the `Unknown` value:

```suggestion
        match *self {
            Self::Unknown(n) => serializer.serialize_newtype_variant("ProcessStatus", "Unknown", n),
            x => serializer.serialize_unit_variant("ProcessStatus", x.to_string()),
        }
```

Reply:
Really better, thanks

Reply:
I think no function taking it as an argument. Maybe I missed it? Which one are you referring to?

Reply:
The two functions for the serialization take the index of the variant as argument unfortunately. The only difference here is for the value of the `newtype_variant`.
See [serialize_unit_variant](https://docs.rs/serde/latest/serde/trait.Serializer.html#tymethod.serialize_unit_variant) and [serialize_newtype_variant](https://docs.rs/serde/latest/serde/trait.Serializer.html#tymethod.serialize_newtype_variant).

Reply:
TIL. Thanks for the explanation. Then I guess it's all good!